src/aliceVision/calibration/checkerDetector.cpp

```diff
@@ -1308,24 +1379,556 @@ bool CheckerDetector::mergeCheckerboards()
+
+        bool keep = true;
+        for (int idCur = 0; idCur < checkers.size(); idCur++)
+        {
+            if (idCur == idRef) continue;
+
+            const CheckerBoard& currentBoard = checkers[idCur].first;
+
+            size_t count_similar = 0;
+            for (int i = 0; i < currentBoard.rows(); i++)
+            {
+                for (int j = 0; j < currentBoard.cols(); j++)
+                {
+                    const IndexT curval = currentBoard(i, j);
+                    if (curval == UndefinedIndexT) continue;
+
+                    if (usedCorners.find(curval) != usedCorners.end())
+                    {
+                        count_similar++;
+                    }
+                }
+            }
+
+            // If the other checker contains almost all the corners of this checkerboard and has a better score
+            double ratio = static_cast<double>(count_similar) / static_cast<double>(usedCorners.size());
+            if (ratio > 0.8)
+            {
+                if (checkers[idRef].second > checkers[idCur].second)
+                {
+                    keep = false;
+                    break;
+                }
             }
         }
+
+        if (keep)
+        {
+            toKeep.push_back(idRef);
+        }
     }
-    while (hadMerged);
 
-    //Copy result
+    // Copy result
     _boards.clear();
-    for (auto b : checkers)
+    for (auto id : toKeep)
+    {
+        _boards.push_back(checkers[id].first);
+    }
+
+    return true;
+}
+
+bool CheckerDetector::removeInvalidCheckerboards()
+{
+    std::vector<CheckerBoard> filtered_boards;
+
+    for (auto b : _boards)
     {
-        _boards.push_back(b.first);
+        bool isInvalid = false;
+
+        for (int i = 0; i < b.rows(); i++)
+        {
+            Vec2 sum {0, 0};
+            int count = 0;
+
+            std::vector<Vec2> dirs;
+
+            // Store directions between consecutive corners in row
+            for (int j = 0; j < b.cols() - 1; j++)
+            {
+                IndexT c1 = b(i, j);
+                IndexT c2 = b(i, j + 1);
+
+                if (c1 == UndefinedIndexT || c2 == UndefinedIndexT) continue;
+
+                Vec2 dir = (_corners[c2].center - _corners[c1].center).normalized();
+                
+                dirs.push_back(dir);
+            }
+
+            // Check that angle between two directions is less than PI/4
+            for (Vec2 ref : dirs)
+            {
+                for (Vec2 cur : dirs)
+                {
+                    double angle = acos(ref.dot(cur));
+                    if (std::abs(angle) > boost::math::constants::pi<double>() * 0.25)
+                    {
+                        isInvalid = true;
+                        break;
+                    }
+                }
+
+                if (isInvalid) break;
+            }
+
+            if (isInvalid) break;
+        }
+
+        if (isInvalid) continue;
+
+        for (int j = 0; j < b.cols(); j++)
```

Comment:
same comment as before, the code is similar in the two loops, maybe it can be refactored or at least split in helper functions.